crypto/bls_core.c

```diff
@@ -1,541 +1,503 @@
-            bn_rand(r, RLC_POS, SEC_BITS);
-            bn_add_dig(r, r, 1); 
-            ep_mul_lwnaf(&sigs[i], &sigs[i], r);
-            ep2_mul_lwnaf(&pks[i], (ep2_st*) &pks_input[i], r);      
-        }
+// - use random coefficients for signatures and public keys at the same index to
+// prevent
+//  indices mixup.
+// - optimize the verification by verifying an aggregated signature against an
+// aggregated
+//  public key, and use a top-down recursive verification to find invalid
+//  signatures.
+void bls_batch_verify(const int sigs_len, byte *results, const E2 *pks_input,
+                      const byte *sigs_bytes, const byte *data,
+                      const int data_len, const byte *seed) {
+
+  // initialize results to undefined
+  memset(results, UNDEFINED, sigs_len);
+
+  // build the arrays of G1 and G2 elements to verify
+  E2 *pks = (E2 *)malloc(sigs_len * sizeof(E2));
+  if (!pks)
+    return;
+  E1 *sigs = (E1 *)malloc(sigs_len * sizeof(E1));
+  if (!sigs)
+    goto out_sigs;
+
+  for (int i = 0; i < sigs_len; i++) {
+    // convert the signature points:
+    // - invalid points are stored as infinity points with an invalid result, so
+    // that the tree aggregations remain valid.
+    // - valid points are multiplied by a random scalar (same for public keys at
+    // same index) to make sure a signature at index (i) is verified against the
+    // public key at the same index.
+    int read_ret =
+        E1_read_bytes(&sigs[i], &sigs_bytes[G1_SER_BYTES * i], G1_SER_BYTES);
+    if (read_ret != VALID || !E1_in_G1(&sigs[i])) {
+      // set signature and key to infinity (no effect on the aggregation tree)
+      // and set result to invalid (result won't be overwritten)
+      E2_set_infty(&pks[i]);
+      E1_set_infty(&sigs[i]);
+      results[i] = INVALID;
+    } else {
+      // choose a random non-zero coefficient of at least 128 bits
+      Fr r, one;
+      // r = random, i-th seed is used for i-th signature
+      Fr_set_zero(&r);
+      const int seed_len = SEC_BITS / 8;
+      limbs_from_be_bytes((limb_t *)&r, seed + (seed_len * i),
+                          seed_len); // faster shortcut than Fr_map_bytes
+      // r = random + 1
+      Fr_set_limb(&one, 1);
+      Fr_add(&r, &r, &one);
+      // multiply public key and signature by the same random exponent r
+      E2_mult(&pks[i], &pks_input[i], &r);
+      E1_mult(&sigs[i], &sigs[i], &r);
     }
-
-    // build a binary tree of aggreagtions
-    node* root = build_tree(sigs_len, &pks[0], &sigs[0]);
-    if (!root) goto out;
-
-    // verify the binary tree and fill the results using batch verification
-    bls_batchVerify_tree(root, sigs_len, &results[0], data, data_len);
-    // free the allocated tree 
-    free_tree(root);
-
+  }
+  // build a binary tree of aggregations
+  node *root = build_tree(sigs_len, &pks[0], &sigs[0]);
+  if (!root)
+    goto out;
+
+  E1 h;
+  if (map_to_G1(&h, data, data_len) != VALID) {
+    goto out;
```

Comment:
if we goto to out then we miss deallocation of root

Reply:
good catch!

Reply:
https://github.com/onflow/flow-go/pull/4817/commits/0587bc6e72ae1ee226bfdcd065236c77c4e0a439